src/main/scala/org/ergoplatform/modifiers/mempool/ErgoTransaction.scala

```diff
@@ -46,48 +70,69 @@ case class ErgoTransaction(override val inputs: IndexedSeq[Input],
     * to accumulate further validation results
     */
   def validateStateless: ValidationResult = {
-    accumulateErrors
+    failFast
+      .demand(outputCandidates.nonEmpty, s"No outputs in transaction $toString")
       .demand(inputs.nonEmpty, s"No inputs in transaction $toString")
       .demand(inputs.size <= Short.MaxValue, s"Too many inputs in transaction $toString")
       .demand(outputCandidates.size <= Short.MaxValue, s"Too many outputCandidates in transaction $toString")
+      .demand(outputCandidates.forall(_.value >= 0), s"Transaction has an output with negative amount $toString")
+      .demand(Try(outputCandidates.map(_.value).reduce(Math.addExact(_, _))).isSuccess, s"Overflow in outputs in $toString")
+      .demand(outAssetsOpt.isSuccess, s"Asset rules violated in $toString: ${outAssetsOpt.failed.get.getMessage}")
       .result
   }
 
   /**
-    * @return total coimputation cost
+    * @return total computation cost
     */
   def statefulValidity(boxesToSpend: IndexedSeq[ErgoBox], blockchainState: ErgoStateContext): Try[Long] = Try {
     require(boxesToSpend.size == inputs.size, s"boxesToSpend.size ${boxesToSpend.size} != inputs.size ${inputs.size}")
 
-    val lastUtxoDigest = AvlTreeData(blockchainState.digest, ErgoBox.BoxId.size)
+    lazy val lastUtxoDigest = AvlTreeData(blockchainState.digest, ErgoBox.BoxId.size)
 
-    val txCost = boxesToSpend.zipWithIndex.foldLeft(0L) { case (accCost, (box, idx)) =>
+    lazy val txCost = boxesToSpend.zipWithIndex.foldLeft(0L) { case (accCost, (box, idx)) =>
       val input = inputs(idx)
-
-      assert(box.id sameElements input.boxId)
+      require(util.Arrays.equals(box.id, input.boxId))
 
       val proof = input.spendingProof
-
       val proverExtension = inputs(idx).spendingProof.extension
-
       val context =
         ErgoLikeContext(blockchainState.height, lastUtxoDigest, boxesToSpend, this, box, proverExtension)
 
       val scriptCost: Long = verifier.verify(box.proposition, context, proof, messageToSign) match {
         case Success((res, cost)) =>
-          if (!res) {
-            throw new Exception(s"Validation failed for input #$idx of tx ${toString}")
-          } else {
-            cost
-          }
+          if (!res) throw new Exception(s"Validation failed for input #$idx of tx $toString") else cost
         case Failure(e) =>
           log.warn(s"Invalid transaction $toString: ", e)
           throw e
       }
       accCost + scriptCost
     }
-    txCost
-  }
+
+    lazy val inputSum = boxesToSpend.map(_.value).reduce(Math.addExact(_, _))
+    lazy val outputSum = outputCandidates.map(_.value).reduce(Math.addExact(_, _))
+
+    def checkAssetPreservationRules = {
```

Comment:
May be better to have `Try` return type here

Reply:
@mike-aksarin is developing a wrapper in Validator for Try() and an arbitrary code block (to be wrapped in Try inside)